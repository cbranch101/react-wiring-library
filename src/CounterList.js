import React, {useState} from 'react'
import ReactDOM from 'react-dom'

const RenderInBody = ({render}) => {
  return ReactDOM.createPortal(render(), document.body)
}

const Counter = ({type, setCount, name, count, increment, decrement}) => {
  const renderType = () => {
    if (type === 'reset') {
      return <button onClick={() => setCount(() => 0)}>Reset</button>
    }
    if (type === 'addTen') {
      return (
        <button onClick={() => setCount(prevCount => prevCount + 10)}>
          Add Ten
        </button>
      )
    }
    return null
  }
  return (
    <li data-type={type} data-testid={`counter_${name}`}>
      <span data-testid="name">{`${name}: `}</span>
      <span data-testid="count">{count}</span>
      <button onClick={increment}>+</button>
      <button onClick={decrement}>-</button>
      {renderType()}
    </li>
  )
}

const List = ({listName, counters, renderCounter}) => {
  return (
    <div data-testid="list">
      <span data-testid="list-name">{listName}</span>
      <ul>{counters.map(counter => renderCounter({counter}))}</ul>
    </div>
  )
}

const Summary = ({counters}) => {
  return (
    <RenderInBody
      render={() => {
        return (
          <span data-testid="summary">
            {`Total value of counters: ${counters.reduce(
              (memo, counter) => memo + counter.count,
              0,
            )}`}
          </span>
        )
      }}
    />
  )
}

export default ({listName, counters}) => {
  const [isShown, setIsShown] = useState(false)
  const [counterValuesByName, setCounterValuesByName] = useState({})
  const updateCountForName = (name, func) => {
    setCounterValuesByName(prevCounterValuesByName => {
      const prevCount = prevCounterValuesByName[name] || 0
      return {
        ...prevCounterValuesByName,
        [name]: func(prevCount),
      }
    })
  }

  const increment = name => updateCountForName(name, prevCount => prevCount + 1)
  const decrement = name => updateCountForName(name, prevCount => prevCount - 1)

  const mappedCounters = counters.map(counter => ({
    ...counter,
    count: counterValuesByName[counter.name] || 0,
  }))
  return (
    <div data-testid="counter-container">
      {isShown ? (
        <List
          listName={listName}
          counters={mappedCounters}
          renderCounter={({counter}) => {
            const {name, count, type} = counter
            return (
              <Counter
                type={type}
                key={name}
                name={name}
                count={count}
                increment={() => increment(name)}
                decrement={() => decrement(name)}
                setCount={func => updateCountForName(name, func)}
              />
            )
          }}
        />
      ) : (
        <button onClick={() => setIsShown(true)}>Show</button>
      )}
      <Summary counters={mappedCounters} />
    </div>
  )
}
