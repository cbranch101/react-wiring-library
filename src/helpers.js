import {within} from 'react-testing-library'
import {getAllFunctions} from './getRenderHandler'

export const combine = (...strings) => {
  return strings.filter(string => string !== undefined).join('\n')
}

export const addAllCustomFunctions = (
  val,
  customFunctions = {},
  customQueries = {},
) => {
  const returnedFromWithin = within(val)
  const {
    global: getGlobalFunctions = () => ({}),
    withinElement: getWithinElementFunctions = functions => functions,
  } = customFunctions
  const globalFunctions = getGlobalFunctions(returnedFromWithin)
  return getAllFunctions(
    {
      ...returnedFromWithin,
      container: val,
    },
    globalFunctions,
    getWithinElementFunctions,
    customQueries,
  )
}
