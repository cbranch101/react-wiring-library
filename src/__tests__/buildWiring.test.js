import {cleanup} from 'react-testing-library'
import React from 'react'
import {combine} from '../helpers'
import buildWiring from '../index'
import CounterList from '../CounterList'

afterEach(() => {
  jest.clearAllMocks()
  cleanup()
})

const fixture = (
  <CounterList
    counters={[
      {
        name: 'first',
        type: 'default',
      },
      {
        name: 'second',
        type: 'reset',
      },
      {
        name: 'third',
        type: 'addTen',
      },
    ]}
    listName="Some counters"
  />
)

const wiring = {
  extend: (val, {findCounterContainer}) => ({
    findCounterList: async () => {
      const {findCounterList, counterContainer} = await findCounterContainer()
      const returned = await findCounterList()
      return {...returned, counterContainer}
    },
  }),
  children: {
    counterContainer: {
      findValue: 'counter-container',
      extend: (val, {findShowButton, findCounterList}) => ({
        findCounterList: async () => {
          const {click} = await findShowButton()
          click()
          return findCounterList()
        },
      }),
      serialize: (val, {counterListString, summaryString}) => {
        return combine(counterListString, summaryString)
      },
      children: {
        showButton: {
          findValue: 'Show',
          findType: 'text',
          extend: (val, {clickElement}) => ({
            click: () => clickElement(val),
          }),
          serialize: () => '[Show]',
        },
        summary: {
          findValue: new RegExp('^Total value of counters'),
          findType: 'text',
          shouldFindInBaseElement: true,
          serialize: val => val.textContent,
        },
        counterList: {
          findValue: 'list',
          serialize: (val, {getTextContents, counterStrings}) => {
            const [listName] = getTextContents(['list-name'])
            return combine(listName, ...counterStrings)
          },
          children: {
            counter: {
              isMultiple: true,
              findValue: new RegExp('^counter'),
              getCurrentType: val => {
                const type = val.getAttribute('data-type')
                return type === 'default' ? undefined : type
              },
              types: {
                addTen: {
                  serialize: (val, {addTenButtonString}, baseString) => {
                    return `${baseString} ${addTenButtonString}`
                  },
                  extend: (val, {findAddTenButton}) => ({
                    addTen: async () => {
                      const {click} = await findAddTenButton()
                      click()
                    },
                  }),
                  children: {
                    addTenButton: {
                      findValue: 'Add Ten',
                      findType: 'text',
                      extend: (val, {clickElement}) => ({
                        click: () => clickElement(val),
                      }),
                      serialize: () => '[Add ten]',
                    },
                  },
                },
                reset: {
                  serialize: (val, {resetButtonString}, baseString) => {
                    return `${baseString} ${resetButtonString}`
                  },
                  extend: (val, {findResetButton}) => ({
                    reset: async () => {
                      const {click} = await findResetButton()
                      click()
                    },
                  }),
                  children: {
                    resetButton: {
                      findValue: 'Reset',
                      findType: 'text',
                      extend: (val, {clickElement}) => ({
                        click: () => clickElement(val),
                      }),
                      serialize: () => '[Reset]',
                    },
                  },
                },
              },
              extend: (val, {clickText}) => {
                return {
                  increase: () => clickText('+'),
                  decrease: () => clickText('-'),
                }
              },
              serialize: (val, {getTextContents}) => {
                const [name, count] = getTextContents(['name', 'count'])
                return `-${name} ${count}`
              },
            },
          },
        },
      },
    },
  },
}

describe('buildWiring helper', () => {
  test('should correctly handler serializers', async () => {
    const getRender = buildWiring(wiring)
    const render = getRender(['counterContainer'])
    const {findCounterList} = render(fixture)
    const {counterContainer, findCounter} = await findCounterList()
    const assertFirstRow = async () => {
      const {increase, decrease} = await findCounter({index: 0})
      expect(counterContainer).toMatchSnapshot(
        'correctly serializes after calling buildWiring',
      )
      increase()
      expect(counterContainer).toMatchSnapshot('counter increases on first row')
      decrease()
      expect(counterContainer).toMatchSnapshot('counter decreases on first row')
    }

    await assertFirstRow()

    const assertSecondRow = async () => {
      const {increase, reset} = await findCounter({
        filter: (counter, testId) => testId === 'counter_second',
      })
      increase()
      expect(counterContainer).toMatchSnapshot(
        'counter increases on second row',
      )
      await reset()
      expect(counterContainer).toMatchSnapshot(
        'counter resets to zero on second row',
      )
    }

    await assertSecondRow()

    const assertThirdRow = async () => {
      const {addTen} = await findCounter({index: 2})
      await addTen()
      expect(counterContainer).toMatchSnapshot('10 is added to the third row')
    }
    await assertThirdRow()

    const assertErrors = async () => {
      await expect(findCounter({index: 10})).rejects.toThrow(
        'You tried to find index 10 in findCounter but 3 is the highest index',
      )
      await expect(findCounter({filter: () => false})).rejects.toThrow(
        "the filter function passed into findCounter didn't find anything",
      )
      await expect(findCounter({filter: () => true})).rejects.toThrow(
        'the filter function passed into findCounter returned 3 elements, it should only return one',
      )
      await expect(findCounter()).rejects.toThrow(
        'You tried to call findCounter which was set as isMultiple, without providing either an index, or a filter function',
      )
    }

    await assertErrors()
  })
  test('should be possible to call manually call serialize', async () => {
    jest.spyOn(console, 'log')
    const getRender = buildWiring(wiring)
    const render = getRender(['counterContainer'])
    const {findCounterList, serialize} = render(fixture)
    const {counterContainer} = await findCounterList()
    serialize(counterContainer)
    expect(console.log.mock.calls[0][0]).toMatchSnapshot(
      'initial render passed into console.log',
    )
    expect(() => serialize({foo: 'bar'})).toThrow(
      "Object can't be serialzied,  make sure it's defined in wiring",
    )
  })
})
