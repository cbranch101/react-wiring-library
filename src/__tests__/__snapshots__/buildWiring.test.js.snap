// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildWiring helper should be possible to call manually call serialize: initial render passed into console.log 1`] = `
"Some counters
-first:  0
-second:  0 [Reset]
-third:  0 [Add ten]
Total value of counters: 0"
`;

exports[`buildWiring helper should correctly handler serializers: 10 is added to the third row 1`] = `
Some counters
-first:  0
-second:  0 [Reset]
-third:  10 [Add ten]
Total value of counters: 10
`;

exports[`buildWiring helper should correctly handler serializers: correctly serializes after calling buildWiring 1`] = `
Some counters
-first:  0
-second:  0 [Reset]
-third:  0 [Add ten]
Total value of counters: 0
`;

exports[`buildWiring helper should correctly handler serializers: counter decreases on first row 1`] = `
Some counters
-first:  0
-second:  0 [Reset]
-third:  0 [Add ten]
Total value of counters: 0
`;

exports[`buildWiring helper should correctly handler serializers: counter increases on first row 1`] = `
Some counters
-first:  1
-second:  0 [Reset]
-third:  0 [Add ten]
Total value of counters: 1
`;

exports[`buildWiring helper should correctly handler serializers: counter increases on second row 1`] = `
Some counters
-first:  0
-second:  1 [Reset]
-third:  0 [Add ten]
Total value of counters: 1
`;

exports[`buildWiring helper should correctly handler serializers: counter resets to zero on second row 1`] = `
Some counters
-first:  0
-second:  0 [Reset]
-third:  0 [Add ten]
Total value of counters: 0
`;
